packages/flutter/lib/src/widgets/widget_state.dart

```diff
@@ -519,12 +686,24 @@ class _WidgetStateTextStyle extends WidgetStateTextStyle {
 ///    `WidgetStateProperty`.
 /// {@macro flutter.widgets.WidgetStateProperty.implementations}
 abstract class WidgetStateProperty<T> {
-  /// Returns a value of type `T` that depends on [states].
+  /// This abstract constructor allows extending the class.
   ///
-  /// Widgets like [TextButton] and [ElevatedButton] apply this method to their
-  /// current [WidgetState]s to compute colors and other visual parameters
-  /// at build time.
-  T resolve(Set<WidgetState> states);
+  /// [WidgetStateProperty] is designed as an interface, so this constructor
+  /// is only needed for backward compatibility.
+  WidgetStateProperty();
+
+  /// Creates a property that resolves using a [WidgetStateMap].
+  ///
+  /// {@template flutter.widgets.WidgetStateProperty.fromMap}
+  /// This constructor's [resolve] method finds the first [MapEntry] whose
+  /// key is satisfied by the set of states, and returns its associated value.
+  /// {@endtemplate}
+  ///
+  /// Returns `null` if no keys match, or if [T] is non-nullable,
+  /// the method throws an [ArgumentError].
+  ///
+  /// {@macro flutter.widgets.WidgetState.any}
+  const factory WidgetStateProperty.fromMap(WidgetStateMap<T> map) = _WidgetStateMapper<T>;
```

Comment:
I suspect that this is a common entry point to WidgetState; maybe include the general example in this doc comment as well to give a general overview of how this all works.